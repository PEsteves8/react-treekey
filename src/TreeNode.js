import React from "react";

import TreeNodeContent from "./TreeNodeContent";
import TreeNodeToggle from "./TreeNodeToggle";

class TreeNode extends React.Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
  }

  onClick(e) {
    e.stopPropagation();
    this.props.selectNewNode(this.props.node, e);
  }

  render() {
    const { style } = this.props;
    const isSelectedNode = this.props.selectedNodes.includes(this.props.node);

    let nodeTextStyles = {};

    if (style) {
      nodeTextStyles = { ...style.node.default };

      if (isSelectedNode) {
        nodeTextStyles = { ...nodeTextStyles, ...style.node.selected };
      }
    }
    const isExpanded = this.props.expandedNodes.includes(this.props.node);
    const templates = this.props.templates || {};

    return (
      <li
        onClick={this.onClick}
        className={`${isSelectedNode ? "treeview-selected-node" : ""}`}
      >
        <div style={nodeTextStyles}>
          {this.props.node.children && (
            <TreeNodeToggle
              template={templates.toggle}
              setToggling={this.props.setToggling}
              node={this.props.node}
              isExpanded={isExpanded}
            />
          )}
          <TreeNodeContent
            template={templates.header}
            node={this.props.node}
            iconsEnabled={this.props.iconsEnabled}
          />
        </div>

        {this.props.node.children && isExpanded && (
          <ul style={this.props.style.nestedList}>
            {this.props.node.children.map((item, idx) => {
              return (
                <TreeNode
                  style={this.props.style}
                  node={item}
                  key={idx}
                  selectNewNode={this.props.selectNewNode}
                  setToggling={this.props.setToggling}
                  iconsEnabled={this.props.iconsEnabled}
                  templates={templates}
                  selectedNode={this.props.selectedNode}
                  selectedNodes={this.props.selectedNodes}
                  expandedNodes={this.props.expandedNodes}
                />
              );
            })}
          </ul>
        )}
      </li>
    );
  }
}

export default TreeNode;
