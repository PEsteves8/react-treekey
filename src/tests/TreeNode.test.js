import React from "react";
import TreeNode from "../TreeNode";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

const mockNode = {
  name: "some node",
};

const expandedNodes = [];

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("<TreeNode>", () => {
  test("should match snapshots", () => {
    act(() => {
      render(
        <TreeNode
          node={mockNode}
          expandedNodes={expandedNodes}
          selectedNodes={[]}
          handleToggle={jest.fn()}
        ></TreeNode>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();

    act(() => {
      render(
        <TreeNode
          node={mockNode}
          expandedNodes={expandedNodes}
          selectedNodes={[mockNode]}
          handleToggle={jest.fn()}
        ></TreeNode>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
  });

  test("calls onSelectNode when clicked", () => {
    const selectNewNode = jest.fn();

    act(() => {
      render(
        <TreeNode
          selectNewNode={selectNewNode}
          node={mockNode}
          expandedNodes={expandedNodes}
          selectedNodes={[]}
          handleToggle={jest.fn()}
        ></TreeNode>,
        container
      );
    });

    const el = container.querySelector("li");
    act(() => {
      el.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(selectNewNode).toHaveBeenCalledTimes(1);
  });
});
