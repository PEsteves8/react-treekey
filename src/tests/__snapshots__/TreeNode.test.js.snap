// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeNode> should match snapshots 1`] = `"<li class=\\"\\"><div><span style=\\"padding-left: 7px;\\">some node</span></div></li>"`;

exports[`<TreeNode> should match snapshots 2`] = `"<li class=\\"treeview-selected-node\\"><div><span style=\\"padding-left: 7px;\\">some node</span></div></li>"`;
